googletakeout.cc

```diff
@@ -0,0 +1,416 @@
+// for some reason that probably only a former Google engineer knows,
+// we use "latE7"/"lngE7" here instead of "latitudeE7"/"longitudeE7".
+// +10 points for brevity, but -100 points for inconsistency.
+#define LATE7 "latE7"
+#define LONE7 "lngE7"
+
+const QList<QString> month_names{
+  "JANUARY", "FEBRUARY", "MARCH", "APRIL", "MAY", "JUNE", "JULY",
+  "AUGUST", "SEPTEMBER", "OCTOBER", "NOVEMBER", "DECEMBER"
+};
+
+static void _fatal(const QString& message) {
+  fatal(FatalMsg() << MYNAME << ": " << message);
+}
+
+static void _warning(const QString& message) {
+  Warning() << MYNAME << ": " << message;
+}
+
+static void title_case(QString& title)
+{
+  bool new_word = true;
+  for (int i = 0; i < title.size(); ++ i) {
+    if (title[i] == '_' || title[i] == ' ') {
+      new_word = true;
+      if (title[i] == '_') {
+        title[i] = ' ';
+      }
+    } else if (new_word) {
+      new_word = false;
+      title[i] = title[i].toUpper();
+    } else {
+      title[i] = title[i].toLower();
+    }
+  }
+}
+
+void
+GoogleTakeoutFormat::rd_init(const QString& fname) {
+  Debug(1) << "rd_init(" << fname << ")";
+  inputStream = GoogleTakeoutInputStream(fname);
+}
+
+void
+GoogleTakeoutFormat::read()
+{
+  int items = 0;
+  int points = 0;
+  int place_visits = 0;
+  int activity_segments = 0;
+  QJsonValue iterator = inputStream.next();
+
+  for (; !iterator.isNull(); iterator = inputStream.next()) {
+    ++ items;
+    /*
+     * A timelineObject is stored in a single-element dictionary.
+     * The key will be either a "placeVisit" (waypoint) or
+     * "activitySegment" (movement), and the value is another dictionary
+     * containing the timelineObject's details.
+     *
+     */
+    const QJsonObject timelineObjectContainer = iterator.toObject();
+    int len = timelineObjectContainer.size();
+    if (len != 1) {
+      _fatal(
+        QString("expected a single key dict, got ") + QString::number(len) +
+        " keys"
+      );
+    }
+    const QJsonObject::const_iterator timelineObjectIterator =
+      timelineObjectContainer.constBegin();
+    const QString& timelineObjectType = timelineObjectIterator.key();
+    const QJsonObject& timelineObjectDetail =
+      timelineObjectIterator.value().toObject();
+    if (timelineObjectType == PLACE_VISIT) {
+      add_place_visit(timelineObjectDetail);
+      ++ place_visits;
+      ++ points;
+    } else if (timelineObjectType == ACTIVITY_SEGMENT) {
+      points += add_activity_segment(timelineObjectDetail);
+      ++ activity_segments;
+    } else {
+      _fatal(
+        QString("unknown timeline object type \"") + timelineObjectType +
+        "\""
+      );
+    }
+  }
+  Debug(1) << MYNAME << ": Processed " << items << " items: " <<
+    place_visits << " " PLACE_VISIT << ", " << activity_segments <<
+    " " ACTIVITY_SEGMENT << " (" << points << " points total)";
+}
+
+/* create a waypoint from late7/lone7 and optional metadata */
+Waypoint*
+GoogleTakeoutFormat::_waypoint(
```

Comment:
warning: method '_waypoint' can be made static [readability-convert-member-functions-to-static]